web/modules/custom/liiweb_api/src/Controller/LiiWebEntityResource.php

```diff
@@ -0,0 +1,296 @@
+use Symfony\Component\HttpKernel\Exception\HttpException;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+class LiiWebEntityResource extends EntityResource {
+
+  const LANGCODE_MAPPING = [
+    'fra' => 'fr',
+    'eng' => 'en',
+  ];
+
+  protected function getRevisionFromFrbrUri($uri) {
+    $this->getNodeFromFrbrUri($uri);
+
+    $result = \Drupal::database()->query("
+    SELECT entity_id, revision_id, langcode
+    FROM node_revision__field_frbr_uri
+    WHERE field_frbr_uri_value = :uri
+    ORDER BY revision_id DESC", [':uri' => $uri])->fetchAssoc();
+
+    if (!empty($result['entity_id'])) {
+      return \Drupal::entityTypeManager()->getStorage('node')->loadRevision($result['revision_id'])->getTranslation($result['langcode']);
+    }
+
+    return NULL;
+  }
+
+  protected function getNodeFromFrbrUri($uri) {
+    $base_uri = explode('/', $uri);
+    array_pop($base_uri);
+    $base_uri = implode('/', $base_uri);
+
+    $node = \Drupal::entityQuery('node')
+      ->condition('type', 'legislation')
+      ->condition('field_frbr_uri', "$base_uri/%", 'LIKE')
+      ->execute();
+
+    if (!empty($node)) {
+      $node = reset($node);
+    }
+
+    return Node::load($node);
+  }
+
+  public function apiCall(ResourceType $resource_type, Request $request, $country, $year, $number, $langcode_year = NULL) {
+    $request_method = $request->getMethod();
+
+    if ($request_method == 'GET') {
+      return $this->get($request);
+    }
+
+    if ($request_method == 'DELETE') {
+      return $this->delete($request, $langcode_year);
+    }
+
+    if (!empty($langcode_year)) {
+      $langcode_year = explode('@', $langcode_year);
+      $date = $langcode_year[1];
+    }
+
+    if ($request_method == 'POST') {
+      /** @var NodeInterface $parsed_entity */
+      $parsed_entity = $this->deserialize($resource_type, $request, JsonApiDocumentTopLevel::class);
+      $create_revision = FALSE;
+      // Try to load the node.
+      $node = $this->getNodeFromFrbrUri($request->getRequestUri());
+      if (empty($node)) {
+        return $this->getResourceResponseError('The requested node does not exist.', 404);
+      }
+
+      $revision = $this->getRevisionFromFrbrUri($request->getRequestUri());
+      if (!empty($revision)) {
+        return $this->getResourceResponseError('Revision already exists.', 400);
+      }
+
+      /** @var \Drupal\Core\Entity\TranslatableInterface $revision */
+      $revision = $this->getRevisionWithPublicationDate($node, $date);
+      // A revision with that creation date does not exist, and the langcode is different from the original language - invalid request
+      if (empty($revision) && $node->language()->getId() != $parsed_entity->language()->getId()) {
+        return $this->getResourceResponseError('Cannot create translations for revisions that do not exist. Please create a revision in the default language for that node with the requested date and try again.', 404);
```

Comment:
Does this mean we need to have an English revision before we can create another language? Can we change the default language on a per-site basis (won't be necessary in the short term, but will in the long term).

Reply:
We've discussed this during the meeting. When a new node is created a language is selected for that node which becomes the "default" node language. When translations are created they are added to that expression, translations cannot exist without original version.

Yes, default language is per site.